File: plugins/wise/engine/src/steps/agent.ts
// `agent` step: build the P6 RunReq from the rendered step and its resolution, start the harness
// child, stream the vendor events into the raw log, and turn the RunRes into step outputs.

import { appendRawLog, writeLog } from "../ledger.ts";
import { effortFor } from "../resolve.ts";
import type {
  AgentStep,
  Effort,
  ExitClass,
  Harness,
  RawEvent,
  Resolved,
  RunReq,
  RunRes,
  Usage,
} from "../types.ts";

export const DEFAULT_STEP_TIMEOUT_MS = 30 * 60 * 1000;
export const VERDICT_MAX = 200;
const LOG_HEAD_BYTES = 2048;

/** First non-empty line of `text`, whitespace collapsed, clipped to `max` chars (E1 verdicts). */
export function headline(text: string, max: number = VERDICT_MAX): string {
  const line = text
    .split("\n")
    .map((l) => l.trim())
    .find((l) => l.length > 0);
  const flat = (line ?? "").replace(/\s+/g, " ");
  return flat.length > max ? flat.slice(0, max - 1) + "…" : flat;
}

/** A started harness child. `startClaude` returns a superset; other adapters give `done` only. */
export type AgentHandle = {
  pid?: number;
  done: Promise<RunRes>;
  /** Mid-run user message (Claude only, M2.6). */
  nudge?: (text: string) => void;
  kill?: (signal?: NodeJS.Signals) => void;
};

export type AgentStarter = (
  harness: Harness,
  req: RunReq,
  onEvent: (e: RawEvent) => void,
) => AgentHandle;

/** Where the child's `wise-engine unit-mcp` finds the daemon (P8): no secrets in argv, all in env. */
export type ChannelConfig = { engineRoot: string; socketPath: string; dataRoot: string };

/** The one MCP server every child loads (D16, D18). `bash engine.sh unit-mcp` picks bun or node. */
export function childMcpConfig(
  channel: ChannelConfig,
  token: string,
): NonNullable<RunReq["mcp_config"]> {
  return {
    mcpServers: {
      "wise-engine": {
        command: "bash",
        args: [`${channel.engineRoot}/engine.sh`, "unit-mcp"],
        env: {
          WISE_STEP_TOKEN: token,
          WISE_ENGINE_SOCKET: channel.socketPath,
          WISE_DATA_ROOT: channel.dataRoot,
        },
      },
    },
  };
}

export type AgentStepInput = {
  runDir: string;
  stepRunId: string;
  /** Already rendered (`renderStep`). */
  step: AgentStep;
  resolved: Resolved;
  cwd: string;
  /** Stored cursor from the previous attempt; honoured only under `resume: unit`. */
  cursor?: unknown;
  stepToken: string;
  starter: AgentStarter;
  defaultTimeoutMs?: number;
  /** When set, the child gets the engine MCP server with this token in its env. */
  channel?: ChannelConfig;
  /** Live-status hook: every vendor event, before it is logged. */
  onEvent?: (e: RawEvent) => void;
};

export type AgentOutcome = {
  exit: ExitClass | "missing_output";
  ok: boolean;
  outputs: Record<string, unknown>;
  verdict: string;
  error?: string;
  usage: Usage;
  cursor?: unknown;
  warnings: string[];
};

export function buildRunReq(input: AgentStepInput): RunReq {
  const { step, resolved } = input;
  const req: RunReq = {
    prompt: step.prompt,
    model: resolved.model,
    cwd: input.cwd,
    mode: step.mode ?? "auto",
    timeout_ms:
      step.timeout !== undefined
        ? step.timeout * 1000
        : (input.defaultTimeoutMs ?? DEFAULT_STEP_TIMEOUT_MS),
    auth: step.auth ?? "subscription",
    step_token: input.stepToken,
  };
  if (resolved.effort !== "" && effortFor(resolved.harness, resolved.effort) !== undefined) {
    req.effort = resolved.effort as Effort;
  }
  if (step.schema !== undefined) req.schema = step.schema;
  if (step.max_turns !== undefined) req.max_turns = step.max_turns;
  // E8: only the `unit` policy resumes a prior session; `fresh` (default) starts clean.
  if (step.resume === "unit" && input.cursor !== undefined) req.resume = input.cursor;
  if (input.channel !== undefined) req.mcp_config = childMcpConfig(input.channel, input.stepToken);
  return req;
}

type Rec = Record<string, unknown>;
const isRec = (v: unknown): v is Rec => typeof v === "object" && v !== null && !Array.isArray(v);

/** Tool names in a vendor event, for the human log (Claude `assistant` shape; others add later). */
function toolNames(parsed: unknown): string[] {
  if (!isRec(parsed) || parsed.type !== "assistant" || !isRec(parsed.message)) return [];
  const content = parsed.message.content;
  if (!Array.isArray(content)) return [];
  return content
    .filter((b): b is Rec => isRec(b) && b.type === "tool_use")
    .map((b) => (typeof b.name === "string" ? b.name : "?"));
}

function excerpt(text: string): string {
  if (text.length <= 2 * LOG_HEAD_BYTES) return text;
  return `${text.slice(0, LOG_HEAD_BYTES)}\n\n[... ${text.length - 2 * LOG_HEAD_BYTES} chars elided ...]\n\n${text.slice(-LOG_HEAD_BYTES)}`;
}

export function humanLog(input: AgentStepInput, req: RunReq, res: RunRes, tools: string[]): string {
  const u = res.usage;
  const lines = [
    `step: ${input.step.id} (${input.stepRunId})`,
    `harness: ${input.resolved.harness} model: ${req.model} effort: ${req.effort ?? "-"} mode: ${req.mode}`,
    `exit: ${res.exit}${res.error ? ` error: ${res.error}` : ""}`,
    `usage: in=${u.input} out=${u.output} cache_read=${u.cache_read} cache_write=${u.cache_write}${u.cost_usd !== undefined ? ` cost_usd=${u.cost_usd}` : ""} pool=${u.pool}`,
    `tools: ${tools.length ? tools.join(", ") : "-"}`,
  ];
  if (res.cursor !== undefined) lines.push(`cursor: ${JSON.stringify(res.cursor)}`);
  if (res.warnings?.length) lines.push(`warnings: ${res.warnings.join("; ")}`);
  lines.push("", "--- text ---", excerpt(res.text));
  if (res.json !== undefined) lines.push("", "--- json ---", excerpt(JSON.stringify(res.json)));
  return lines.join("\n") + "\n";
}

/** Copy the declared output names out of the schema result; a missing name fails the step. */
export function extractOutputs(
  names: readonly string[] | undefined,
  json: unknown,
): { outputs: Record<string, unknown>; missing?: string } {
  const outputs: Record<string, unknown> = {};
  if (!names || names.length === 0) return { outputs };
  if (!isRec(json)) return { outputs, missing: names[0] as string };
  for (const name of names) {
    if (!(name in json) || json[name] === undefined) return { outputs, missing: name };
    outputs[name] = json[name];
  }
  return { outputs };
}

export function outcomeOf(step: AgentStep, res: RunRes): AgentOutcome {
  const base = { usage: res.usage, warnings: res.warnings ?? [] };
  const withCursor = (o: AgentOutcome): AgentOutcome => {
    if (res.cursor !== undefined) o.cursor = res.cursor;
    return o;
  };
  if (res.exit !== "ok") {
    const error = res.error ?? res.exit;
    return withCursor({
      ...base,
      exit: res.exit,
      ok: false,
      outputs: {},
      verdict: headline(`${res.exit}: ${error}`),
      error,
    });
  }
  const { outputs, missing } = extractOutputs(step.outputs, res.json);
  if (missing !== undefined) {
    const error = `schema result lacks ${missing}`;
    return withCursor({
      ...base,
      exit: "missing_output",
      ok: false,
      outputs: {},
      verdict: headline(error),
      error,
    });
  }
  const verdict =
    headline(res.text) || (res.json !== undefined ? headline(JSON.stringify(res.json)) : "ok");
  return withCursor({ ...base, exit: "ok", ok: true, outputs, verdict });
}

/** Start the child; `handle` is live at once (kill / nudge), `outcome` settles when it exits. */
export function startAgentStep(input: AgentStepInput): {
  handle: AgentHandle;
  req: RunReq;
  outcome: Promise<AgentOutcome>;
} {
  const req = buildRunReq(input);
  const tools: string[] = [];
  const onEvent = (e: RawEvent): void => {
    tools.push(...toolNames(e.parsed));
    try {
      input.onEvent?.(e);
    } catch {
      // A status hook failure never fails the step either.
    }
    try {
      appendRawLog(input.runDir, input.step.id, input.stepRunId, e);
    } catch {
      // A log write failure never fails the step.
    }
  };
  const handle = input.starter(input.resolved.harness, req, onEvent);
  const outcome = handle.done.then((res) => {
    try {
      writeLog(input.runDir, input.step.id, input.stepRunId, humanLog(input, req, res, tools));
    } catch {
      // Same: the ledger's step record is the source of truth.
    }
    return outcomeOf(input.step, res);
  });
  return { handle, req, outcome };
}

File: plugins/wise/engine/wise_engine/steps/agent.py
from __future__ import annotations

import asyncio
import os
import re
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from ..adapter_types import AgentHandle
from ..ledger import append_raw_log, write_log
from ..paths import ENGINE_ROOT, PLUGIN_ROOT
from ..render import _json
from ..resolve import effort_for
from ..scheduler import JS_WHITESPACE, number_text

DEFAULT_STEP_TIMEOUT_MS = 30 * 60 * 1000
VERDICT_MAX = 200
LOG_HEAD_BYTES = 2048
Json = dict[str, Any]
_SPACE = re.compile(f"[{re.escape(JS_WHITESPACE)}]+")
INSTRUCTION_NAMES = ("CLAUDE.md", "AGENTS.md")


def project_system_prompt(cwd: str, home: str | None = None) -> str:
    root = Path(cwd).resolve()
    user_home = Path(home or os.environ.get("HOME") or Path.home()).resolve()
    directories = [root, *root.parents]
    if user_home in directories:
        directories = directories[: directories.index(user_home) + 1]
    paths = [user_home / ".claude" / "CLAUDE.md", user_home / ".codex" / "AGENTS.md"]
    paths.extend(
        directory / name for directory in reversed(directories) for name in INSTRUCTION_NAMES
    )
    blocks = []
    seen = set()
    for path in paths:
        resolved = path.resolve()
        if resolved in seen:
            continue
        seen.add(resolved)
        try:
            content = resolved.read_text(encoding="utf-8").strip()
        except OSError:
            continue
        if content:
            blocks.append(f"### {resolved}\n\n{content}")
    loaded = "\n\n".join(blocks) or "(No applicable files found at child startup.)"
    return f"""# Repository instruction contract

CLAUDE.md and AGENTS.md files are mandatory project instructions for this task,
regardless of which harness is running. Follow all applicable files below. A
file in a deeper directory adds to or overrides broader instructions for work
inside its directory.

Before reading or changing a path below the working directory, check its path
for a closer CLAUDE.md or AGENTS.md and follow it too. If you create any
subagent, teammate, Task, or Agent, pass this entire contract and every
applicable instruction file to it before its task details, and require it to do
the same recursively. Harness defaults do not override these project rules.

## Instructions loaded for {root}

{loaded}"""


def utf16_length(text: str) -> int:
    return len(text.encode("utf-16-le", "surrogatepass")) // 2


def utf16_slice(text: str, start: int = 0, end: int | None = None) -> str:
    encoded = text.encode("utf-16-le", "surrogatepass")
    return encoded[start * 2 : None if end is None else end * 2].decode(
        "utf-16-le", "surrogatepass"
    )


def headline(text: str, maximum: int = VERDICT_MAX) -> str:
    line = next((s.strip(JS_WHITESPACE) for s in text.split("\n") if s.strip(JS_WHITESPACE)), "")
    flat = _SPACE.sub(" ", line)
    return utf16_slice(flat, 0, maximum - 1) + "…" if utf16_length(flat) > maximum else flat


def child_mcp_config(channel: Json, token: str) -> Json:
    engine_root = str(channel.get("engine_root", ENGINE_ROOT))
    return {
        "mcpServers": {
            "wise-engine": {
                "command": sys.executable,
                "args": ["-m", "wise_engine", "unit-mcp"],
                "env": {
                    "PYTHONPATH": engine_root,
                    "WISE_STEP_TOKEN": token,
                    "WISE_ENGINE_SOCKET": channel["socket_path"],
                    "WISE_DATA_ROOT": channel["data_root"],
                },
            }
        }
    }


def build_run_req(params: Json) -> Json:
    step, resolved = params["step"], params["resolved"]
    req = dict(
        prompt=step["prompt"],
        model=resolved["model"],
        cwd=params["cwd"],
        mode=step.get("mode", "auto"),
        timeout_ms=step["timeout"] * 1000
        if "timeout" in step
        else params.get("default_timeout_ms", DEFAULT_STEP_TIMEOUT_MS),
        auth=step.get("auth", "subscription"),
        step_token=params["step_token"],
        add_dirs=[str(params["run_dir"]), str(PLUGIN_ROOT), *params.get("add_dirs", [])],
        system=project_system_prompt(params["cwd"]),
    )
    for source, target in [
        ("allowed_tools", "allowed_tools"),
        ("mcp", "mcp_policy"),
        ("schema", "schema"),
        ("max_turns", "max_turns"),
    ]:
        if source in step:
            req[target] = step[source]
    if (
        resolved.get("effort", "") != ""
        and effort_for(resolved["harness"], resolved["effort"]) is not None
    ):
        req["effort"] = resolved["effort"]
    if step.get("resume") == "unit" and "cursor" in params:
        req["resume"] = params["cursor"]
    if "channel" in params:
        req["mcp_config"] = child_mcp_config(params["channel"], params["step_token"])
    return req


def _tool_names(parsed: Any) -> list[str]:
    if not isinstance(parsed, dict) or parsed.get("type") != "assistant":
        return []
    message = parsed.get("message")
    if not isinstance(message, dict) or not isinstance(message.get("content"), list):
        return []
    return [
        block["name"] if isinstance(block.get("name"), str) else "?"
        for block in message["content"]
        if isinstance(block, dict) and block.get("type") == "tool_use"
    ]


def _excerpt(text: str) -> str:
    length = utf16_length(text)
    if length <= 2 * LOG_HEAD_BYTES:
        return text
    return f"{utf16_slice(text, 0, LOG_HEAD_BYTES)}\n\n[... {length - 2 * LOG_HEAD_BYTES} chars elided ...]\n\n{utf16_slice(text, -LOG_HEAD_BYTES)}"


def human_log(params: Json, req: Json, res: Json, tools: list[str]) -> str:
    usage = res["usage"]
    lines = [
        f"step: {params['step']['id']} ({params['step_run_id']})",
        f"harness: {params['resolved']['harness']} model: {req['model']} effort: {req.get('effort', '-')} mode: {req['mode']}",
        f"exit: {res['exit']}" + (f" error: {res['error']}" if res.get("error") else ""),
        f"usage: in={number_text(usage['input'])} out={number_text(usage['output'])} cache_read={number_text(usage['cache_read'])} cache_write={number_text(usage['cache_write'])}"
        + (f" cost_usd={number_text(usage['cost_usd'])}" if "cost_usd" in usage else "")
        + f" pool={usage['pool']}",
        f"tools: {', '.join(tools) if tools else '-'}",
    ]
    if "cursor" in res:
        lines.append(f"cursor: {_json(res['cursor'])}")
    if res.get("warnings"):
        lines.append(f"warnings: {'; '.join(res['warnings'])}")
    lines.extend(["", "--- text ---", _excerpt(res["text"])])
    if "json" in res:
        lines.extend(["", "--- json ---", _excerpt(_json(res["json"]))])
    return "\n".join(lines) + "\n"


def extract_outputs(names: list[str] | None, value: Any) -> Json:
    outputs: Json = {}
    for name in names or []:
        if not isinstance(value, dict) or name not in value:
            return dict(outputs=outputs, missing=name)
        outputs[name] = value[name]
    return {"outputs": outputs}


def outcome_of(step: Json, res: Json) -> Json:
    base = {"usage": res["usage"], "warnings": res.get("warnings", [])}
    for key in ("cursor", "json"):
        if key in res:
            base[key] = res[key]
    if res["exit"] != "ok":
        error = res.get("error", res["exit"])
        return {
            **base,
            "exit": res["exit"],
            "ok": False,
            "outputs": {},
            "verdict": headline(f"{res['exit']}: {error}"),
            "error": error,
        }
    extracted = extract_outputs(step.get("outputs"), res.get("json"))
    if "missing" in extracted:
        error = f"schema result lacks {extracted['missing']}"
        return {
            **base,
            "exit": "missing_output",
            "ok": False,
            "outputs": {},
            "verdict": headline(error),
            "error": error,
        }
    verdict = headline(res["text"]) or (headline(_json(res["json"])) if "json" in res else "ok")
    return {**base, "exit": "ok", "ok": True, "outputs": extracted["outputs"], "verdict": verdict}


@dataclass
class StartedAgent:
    handle: AgentHandle
    req: Json
    outcome: asyncio.Task[Json]


async def start_agent_step(params: Json) -> StartedAgent:
    req = build_run_req(params)
    tools = []

    def on_event(event: Json) -> None:
        tools.extend(_tool_names(event.get("parsed")))
        try:
            if params.get("on_event"):
                params["on_event"](event)
        except Exception:
            pass
        try:
            append_raw_log(params["run_dir"], params["step"]["id"], params["step_run_id"], event)
        except Exception:
            pass

    handle = await params["starter"](params["resolved"]["harness"], req, on_event)

    async def finish() -> Json:
        result = await handle.done
        try:
            write_log(
                params["run_dir"],
                params["step"]["id"],
                params["step_run_id"],
                human_log(params, req, result, tools),
            )
        except Exception:
            pass
        return outcome_of(params["step"], result)

    return StartedAgent(handle, req, asyncio.create_task(finish()))
